link-grammar/fast-match.c

```diff
@@ -276,64 +334,262 @@ fast_matcher_t* alloc_fast_matcher(const Sentence sent)
+ * link-parser < file.batch | grep match_stats: | sort | uniq -c
+ */
+static void match_stats(Connector *c1, Connector *c2)
+{
+	if (NULL == c1) printf("match_stats: cache\n");
+	if (NULL == c2) return;
+	if ((1 == c1->uc_start) && (1 == c2->uc_start) &&
+	    (c1->string[0] == c2->string[0]))
+	{
+		printf("match_stats: h/d mismatch\n");
+	}
+
+	if (0 == c1->lc_start) printf("match_stats: no lc (c1)\n");
+	if (0 == c2->lc_start) printf("match_stats: no lc (c2)\n");
+
+	if (string_set_cmp(c1->string, c2->string)) printf("match_stats: same\n");
+
+	const char *a = &c1->string[c1->lc_start];
+	const char *b = &c2->string[c2->lc_start];
+	do
+	{
+		if (*a != *b && (*a != '*') && (*b != '*')) printf("match_stats: lc false\n");
+		a++;
+		b++;
+	} while (*a != '\0' && *b != '\0');
+	printf("match_stats: lc true\n");
+}
+#else
+#define match_stats(a, b)
+#endif
+
+#ifdef DEBUG
+#undef N
+#define N(c) (c?c->string:"")
+
+/**
+ * Print the match list, including connector match indications.
+ * Usage: link-parser -v=5 [-debug=print_match_list]
+ * Output format:
+ * MATCH_NODE list_id:  lw>lc   [=]   left<w>right   [=]    rc<rw
+ *
+ * Regretfully this version doesn't indicate which match shortcut have been
+ * used, and which nodes are from mr or ml or both (the full print version
+ * clutters the source code very much, as it needs to be inserted in plenty
+ * of places.)
+ *
+ * FIXME It is not clear to me why these printout are intermixed by stderr
+ * printouts, as prt_error() flushes stdout. For example:
+ *
+ * MATCH_NODE    49: 02>O*t       =        Os<04>                     <06
+ * link-grammar: Info: Total count with 0 null links:   4
+ * ++++Counted parses                          0.00 seconds
+ * MATCH_NODE    58: 00>RW        =        RW<05>                     <06
+ */
+static void print_match_list(int id, Match_node *m, int w,
+                             Connector *lc, int lw,
+                             Connector *rc, int rw)
+{
+	if (verbosity < 5) return;
+
+	for (; m != NULL; m = m->next)
+	{
+		Disjunct *d = m->d;
+
+		printf("MATCH_NODE %5d: %02d>%-9s %c %9s<%02d>%-9s %c %9s<%02d\n",
+		       id, lw , N(lc), d->match_left ? '=': ' ',
+		       N(d->left), w, N(d->right),
+		       d->match_right? '=' : ' ', N(rc), rw);
+	}
+}
+#else
+#define print_match_list(...)
+#endif
+
+/**
+ * Compare two connectors, utilizing features of the match lists.
+ * This function uses shortcuts to speed up the comparison.
+ * Especially, we know that the uc parts of the connectors are the same,
+ * because we fetch the matching lists according to the uc part or the
+ * connectors to be matched. So the uc parts are not checked here. The
+ * head/dependent indicators are in the caller function, and only when
+ * connectors match here, to save CPU when the connectors don't match
+ * otherwise. This is because h/d mismatch is rare.
+ * FIXME: Use connector enumeration.
+ */
+static bool easy_match_list(Connector *c1, Connector *c2)
```

Comment:
could you rename this to "match_lower_case", and change the first two setneces of description above to say "Compare only the lower-case parts of two conectors. When this function is called, we already know that the upper-case parts of the connectors match" ... err, never mind, I will make this change myself right now ...